cpp/src/arrow/filesystem/azurefs_test.cc

```diff
@@ -0,0 +1,532 @@
+    fileSystemClient.CreateIfNotExists();
+    fileSystemClient = gen2Client_->GetFileSystemClient("empty-container");
+    fileSystemClient.CreateIfNotExists();
+    auto fileClient =
+        std::make_shared<Azure::Storage::Files::DataLake::DataLakeFileClient>(
+            options_.account_blob_url + "container/somefile",
+            options_.storage_credentials_provider);
+    std::string s = "some data";
+    fileClient->UploadFrom(const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(&s[0])),
+                           s.size());
+  }
+  void TearDown() override {
+    auto containers = gen2Client_->ListFileSystems();
+    for (auto c : containers.FileSystems) {
+      auto fileSystemClient = gen2Client_->GetFileSystemClient(c.Name);
+      fileSystemClient.DeleteIfExists();
+    }
+  }
+  void AssertObjectContents(
+      Azure::Storage::Files::DataLake::DataLakeServiceClient* client,
+      const std::string& container, const std::string& path_to_file,
+      const std::string& expected) {
+    auto pathClient_ =
+        std::make_shared<Azure::Storage::Files::DataLake::DataLakePathClient>(
+            client->GetUrl() + container + "/" + path_to_file,
+            options_.storage_credentials_provider);
+    auto size = pathClient_->GetProperties().Value.FileSize;
+    if (size == 0) {
+      return;
+    }
+    auto buf = AllocateResizableBuffer(size, fs_->io_context().pool());
+    Azure::Storage::Blobs::DownloadBlobToOptions downloadOptions;
+    Azure::Core::Http::HttpRange range;
+    range.Offset = 0;
+    range.Length = size;
+    downloadOptions.Range = Azure::Nullable<Azure::Core::Http::HttpRange>(range);
+    auto fileClient_ =
+        std::make_shared<Azure::Storage::Files::DataLake::DataLakeFileClient>(
+            client->GetUrl() + container + "/" + path_to_file,
+            options_.storage_credentials_provider);
+    auto result = fileClient_
+                      ->DownloadTo(reinterpret_cast<uint8_t*>(buf->get()->mutable_data()),
+                                   size, downloadOptions)
+                      .Value;
+    buf->get()->Equals(
+        Buffer(const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(&expected[0])),
+               expected.size()));
+  }
+};
+
+TEST_F(TestAzureFileSystem, CreateDir) {
+  // New container
+  AssertFileInfo(fs_.get(), "container3", FileType::NotFound);
+  ASSERT_OK(fs_->CreateDir("container3"));
+  AssertFileInfo(fs_.get(), "container3", FileType::Directory);
+
+  // Existing container
+  ASSERT_OK(fs_->CreateDir("container"));
+  AssertFileInfo(fs_.get(), "container", FileType::Directory);
+
+  ASSERT_RAISES(IOError, fs_->CreateDir(""));
+
+  // Existing "file", should fail
+  ASSERT_RAISES(IOError, fs_->CreateDir("container/somefile"));
+
+  // directory, false
+  ASSERT_RAISES(IOError, fs_->CreateDir("container/newdir/newsub/newsubsub", false));
+
+  // directory, true
+  ASSERT_RAISES(IOError, fs_->CreateDir("container/newdir/newsub/newsubsub", true));
```

Comment:
This fails because Azurite creates an [Azure Blob type](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blobs-introduction) storage account that doesn't support hierarchical namespace. There is no concept of directories inside a container.